Push update.rdf to support self-hosted updates.

ochameau opened the pull request:
Uses same setup we were using for adbhelper.
It would surely help to avoid hardcoding the updateURL in the install.rdf...

So the xpi should live here:
  https://archive.mozilla.org/pub/labs/devtools/master/devtools.xpi
  The install.rdf in this xpi is now going to refer to the following url:
The update.rdf here:
  https://archive.mozilla.org/pub/labs/devtools/master/update.rdf

We can easily introduce some folders, and duplicate all these devtools.xpi and update.rdf to help supporting old releases. One folder per Firefox release
https://archive.mozilla.org/pub/labs/devtools/{56,57,58,...}/{devtools,update}.rdf
`release.sh` would only update the current nightly, beta and release folders and stop updating all the older.
(Note that s3 doesn't support symlink from what I remember so we have to push the same copy multiple times)

I pushed it to master to verify it works and made a copy over `master-update-demo` (in case auto-updater overrides master in the meantime).

juliandescottes commented on bin/release.sh:
```diff
@@ -48,6 +49,15 @@ S3_BASE_URL="s3://net-mozaws-prod-delivery-contrib$S3_ROOT_PATH"
 
+# Feth the final addon version from install.rdf
+# -E is for using regexp (not only strings) and -o is for printing only the pattern that matches
+VERSION=$(grep -E "em:version=\"([0-9.ab-]+)\"" $SCRIPT_DIR/../install.rdf | grep -oE "([0-9.ab-]+)")
```
would this be enough or are you afraid there could be invalid em:version lines ?

`VERSION=$(grep -E "em:version" $SCRIPT_DIR/../install.rdf | grep -oE "([0-9.ab-]+)")`

ochameau replied:
Yes, it should work. I imagine the best solution would be to have a template instead.

juliandescottes commented on bin/release.sh:
```diff
@@ -48,6 +49,15 @@ S3_BASE_URL="s3://net-mozaws-prod-delivery-contrib$S3_ROOT_PATH"
+echo "VERSION=$VERSION"
+
+# Upload also the update.rdf file to S3
+sed -e "s#@@UPDATE_LINK@@#$ADDON_URL#;s#@@ADDON_VERSION@@#$VERSION#" $SCRIPT_DIR/template-update.rdf > update.rdf
```
why are we using # here?

ochameau replied:
So. This is the power of `sed` the delimiter can be any character!
So here, I'm using `#` instead of the very traditional `/`.
That to avoid confusion with $ADDON_URL, which contains many `/` and would make things crash.

juliandescottes commented on bin/template-update.rdf:
```diff
@@ -0,0 +1,17 @@
+      <Description>
+        <em:id>{ec8030f7-c20a-464f-9b0e-13a3a9e97384}</em:id>
+        <em:minVersion>26.0a1</em:minVersion>
+        <em:maxVersion>27.0</em:maxVersion>
```
26.0a1 to 27.0 seems wrong but I have often seen those compatibility fields being ignored (ie I could install an addon compatible up to ff30 on later versions of ff). 
Tried checking the docs for this (https://developer.mozilla.org/en/docs/Extension_Versioning,_Update_and_Compatibility#Compatibility_Updates) but can't find a clear answer: do we need to set this correctly?

ochameau replied:
Yes, it is just wrong I copy pasted that from adbhelper.
Here is a in-production one that look more correct:
https://archive.mozilla.org/pub/labs/valence/win32/update.rdf

juliandescottes commented on bin/release.sh:
```diff
@@ -28,7 +28,8 @@ export AMO_SECRET=$(curl ${secret_url} | jq ".secret.AMO_SECRET" -r)
 # /!\ this changes install.rdf without commiting it.
 VERSION_SUFFIX=$(date +%Y%m%d%H%M)
-sed -i -E 's/em:version=\"([0-9.ab-]+)\"/em:version=\"\1.'$VERSION_SUFFIX'\"/g' install.rdf
+echo "VERSION_SUFFIX=$VERSION_SUFFIX"
```
Was this logged for debugging? Maybe add a small echo before to say we are starting the release script or something?

ochameau replied:
Yes. I may help if the regexp doesn't crash completely but ends up being wrong.

ochameau commented on bin/release.sh:
```diff
@@ -26,9 +28,15 @@ export AMO_SECRET=$(curl ${secret_url} | jq ".secret.AMO_SECRET" -r)
 # /!\ this changes install.rdf without commiting it.
-VERSION_SUFFIX=$(date +%Y%m%d%H%M)
-sed -i -E 's/em:version=\"([0-9.ab-]+)\"/em:version=\"\1.'$VERSION_SUFFIX'\"/g' install.rdf
+VERSION_SUFFIX=$(date +%Y%m%d.%H%M)
```
I had to change the version suffix to please the add-on manager. 
See the inlined comment.

ochameau commented:
I addressed your comments. I pushed again to `master` and archived a copy on 
https://github.com/ochameau/ff-dt/tree/master-update-demo
It spawn this task: 
https://tools.taskcluster.net/task-group-inspector/#/Rip4UWh_Te6boV93shGVkw/UR3vQBR5QgCIR7Mf8XPK-Q?_k=ggs3gn
```
upload: ./devtools-signed.xpi to s3://net-mozaws-prod-delivery-contrib/pub/labs/devtools/master/devtools.xpi
VERSION=44.0a1.20170603.1949
upload: ./update.rdf to s3://net-mozaws-prod-delivery-contrib/pub/labs/devtools/master/update.rdf
```
The update.rdf has better version range:
https://archive.mozilla.org/pub/labs/devtools/master/update.rdf

I'll let you merge this whenever you have time to review it again (ideally on Monday ;)).
